File: Pythoncode/96.不同的二叉搜索树.py
class Solution:
    def __init__(self):
        self.n2numTrees = {}
        self.n2numTrees[0] = 1

    def numTrees(self, n: int) -> int:
        if n in self.n2numTrees.keys():
           return self.n2numTrees[n]
        s = 0
        for i in range(n):
            s += self.numTrees(i) * self.numTrees(n-i-1)
        self.n2numTrees[n] = s
        return s

File: Cppcode/96.不同的二叉搜索树.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int numTrees(int n) {
        vector<int> dp;
        dp.push_back(1);
        for (auto i = 1; i <= n; i++)
        {
            auto a = 0;
            for (auto j = 0; j < i; j++)
            {
                a += dp[j] * dp[i-j-1];
            }
            dp.push_back(a);
        }
        return dp.back();
    }
};